ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
+		Check_Type(rb_server, T_ARRAY);
+
+		rb_backend = rb_ary_entry(rb_server, 0);
+		Check_Type(rb_backend, T_SYMBOL);
+
+		if (SYM2ID(rb_backend) == id_tcp) {
+			VALUE rb_hostname = rb_ary_entry(rb_server, 1);
+			VALUE rb_port = rb_ary_entry(rb_server, 2);
+
+			Check_Type(rb_server, T_STRING);
+			Check_Type(rb_port, T_FIXNUM);
+
+			// TODO: add weight
+			//VALUE rb_weight = rb_ary_entry(rb_server, 3);
+			rc = memcached_server_add(mc, StringValueCStr(rb_hostname), NUM2INT(rb_port));
+			rb_memcached_error_check(rc);
+		}
+		else if (SYM2ID(rb_backend) == id_socket) {
+			VALUE rb_fd = rb_ary_entry(rb_server, 1);
+			Check_Type(rb_fd, T_STRING);
+			rc = memcached_server_add_unix_socket(mc, StringValueCStr(rb_fd));
+			rb_memcached_error_check(rc);
+		}
+	}
+
+	// Verify keys by default
+	rc = memcached_behavior_set(mc, MEMCACHED_BEHAVIOR_VERIFY_KEY, true);
+	rb_memcached_error_check(rc);
+
+	return rb_mc;
 }
 
 static memcached_return_t
 iterate_server_function(const memcached_st *ptr, const memcached_instance_st * server, void *context)
 {
-  VALUE server_list = (VALUE) context;
-
-  VALUE rb_server = rb_obj_alloc(rb_cServer);
-
-  rb_ivar_set(rb_server, id_ivar_hostname, rb_str_new2(memcached_server_name(server)));
-  rb_ivar_set(rb_server, id_ivar_port, UINT2NUM(memcached_server_port(server)));
-  // TODO weight
+	VALUE rb_server_list = (VALUE)context;
+	VALUE rb_server = rb_funcall(rb_cServer, rb_intern("new"), 2, 
+			rb_str_new2(memcached_server_name(server)),
+			UINT2NUM(memcached_server_port(server)));
 
-  rb_ary_push(server_list, rb_server);
-  return MEMCACHED_SUCCESS;
+	rb_ary_push(rb_server_list, rb_server);
+	return MEMCACHED_SUCCESS;
 }
 
 static VALUE
 rb_connection_servers(VALUE self)
 {
-  memcached_ctx *ctx = get_ctx(self);
+	memcached_server_fn callbacks[1] = { &iterate_server_function };
+	memcached_st *mc;
 
-  VALUE server_list = rb_ary_new();
+	VALUE rb_result;
 
-  memcached_server_fn callbacks[1];
-  callbacks[0] = iterate_server_function;
+	UnwrapMemcached(self, mc);
+	rb_result = rb_ary_new();
 
-  memcached_server_cursor(ctx->memc, callbacks, (void *)server_list, 1);
-  return server_list;
+	memcached_server_cursor(mc, callbacks, (void *)rb_result, 1);
+	return rb_result;
 }
 
 static VALUE
 rb_connection_flush(VALUE self)
 {
-  memcached_ctx *ctx = get_ctx(self);
+	memcached_st *mc;
+	memcached_return_t rc;
 
-  memcached_return_t rc = memcached_flush(ctx->memc, 0);
-  handle_memcached_return(rc);
+	UnwrapMemcached(self, mc);
+	rc = memcached_flush(mc, 0);
+	rb_memcached_error_check(rc);
 
-  return (rc == MEMCACHED_SUCCESS);
+	return (rc == MEMCACHED_SUCCESS) ? Qtrue : Qfalse;
 }
 
 static VALUE
-rb_connection_set(VALUE self, VALUE key, VALUE value, VALUE ttl, VALUE flags)
+rb_connection_set(VALUE self, VALUE rb_key, VALUE rb_value, VALUE rb_ttl, VALUE rb_flags)
 {
-  memcached_ctx *ctx = get_ctx(self);
-
-  char *mkey = StringValuePtr(key);
-  char *mvalue = StringValuePtr(value);
```

Comment:
Another big change throughout the library: don't use `StringValuePtr`! It expects the strings to be NULL terminated and to have no NULLs. This defeats the whole point of using `libmemcached`! All APIs in memcached take a explicit length for all strings, because Memcached keys and values can have NULLs in them! Particularly important when writing serialized blobs. This would have failed in a very surprising manner.